### ISSUE TITLE
reportInvalidTypeForm false positive for recursive type alias used in annotation within class namespace

### ISSUE THREAD

**Issue by Jemtaly** [microsoft/pyright#11246]
**Describe the bug**

When a type alias defined in a class namespace has a circular reference with another type alias in the same namespace, Pyright incorrectly reports `reportInvalidTypeForm` ("Variable not allowed in type expression") when using that alias in a type annotation.

However, if the type alias does not have a circular reference, the same pattern works without error. This appears to be a false positive.

**Code or Screenshots**

```python
class Namespace1:
    A = list["Namespace1.B"]
    B = list["Namespace1.A"]  # Circular reference

    def f(self) -> A:
        a: Namespace1.A = []  # Error: Variable not allowed in type expression
        ...
        return a

class Namespace2:
    A = list["Namespace2.B"]
    B = int  # No circular reference

    def f(self) -> A:
        a: Namespace2.A = []  # No error
        ...
        return a
```

[Pyright Playground](https://pyright-play.net/?locale=en&code=MYGwhgzhAEByYFsCmEAOZhIIwC4CwAUNMdAILQC80IAlhAC4DaARPMmhtgHQBCzAuoUIloPStTpNWiFOkxYupAUKIkAJkgBm0TQAoISEJoCU0ALQA%2BMvlUjiYHHBkd5i8Y0G27XH8LvEAJyR6AFcAgDtoMBVQSBg2WU4AJhsRcipaBhYElyQk3mUCP2IxKhpw%2BhURDW09AyNTS2tiuwcndjk8tyoPFpEfLj6SINCIqMIgA)

**Expected behavior**

Both `Namespace1.A` and `Namespace2.A` should be recognized as valid type aliases usable in type annotations. The circular reference between type aliases should not cause `A` to be treated as a regular variable.

**VS Code extension or command-line**

Tested on Pyright Playground (appears to be latest version).

**Comment by erictraut** [comment]
I don't consider this a bug. This is a real circular dependency that cannot be resolved in the general case. The name `"Namespace1"` evaluates the (possibly decorated) class `Namespace1`. You don't have a class decorator on your class, but class decorators need to be supported in general. A class decorator needs to be evaluated and applied to the inner (undecorated class) which includes any symbols included within its namespace — including type aliases. For example, if the class decorator accepts a protocol, a type checker needs to determine whether the class matches that protocol. 

You'll need to break this circular dependency in your code if you want this to work with pyright. 

Other type checkers might ignore this circular dependency, but that will create problems in the general case.
